### ISSUE TITLE
`docker ps` takes almost 4 minutes

### ISSUE THREAD

**Issue by tinco** [docker/docker#15084]
Hi, I've got this output:

BUG REPORT INFORMATION
==========
```
root@us-compute-1 ~ # docker -D info
Containers: 46
Images: 2262
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 2354
 Dirperm1 Supported: false
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.13.0-54-generic
Operating System: Ubuntu 14.04.2 LTS
CPUs: 8
Total Memory: 31.36 GiB
Name: us-compute-1
ID: 4T26:DVKS:7MAQ:MQ2C:EEH2:VPOV:EC5G:MJWL:NOEL:ZS3T:D4AM:NMKO
WARNING: No swap limit support
```
```
root@us-compute-1 ~ # docker version
Client version: 1.7.0
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 0baf609
OS/Arch (client): linux/amd64
Server version: 1.7.0
Server API version: 1.19
Go version (server): go1.4.2
Git commit (server): 0baf609
OS/Arch (server): linux/amd64
```

This is a physical machine with no virtualization whatsoever. Linux directly on hardware.
```
root@us-compute-1 ~ # uname -a
Linux us-compute-1 3.13.0-54-generic #91-Ubuntu SMP Tue May 26 19:15:08 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

# Describe the results you received:

```
root@us-compute-1 ~ # time docker ps
CONTAINER ID        IMAGE                                                      COMMAND                CREATED             STATUS              PORTS                                      NAMES
c3d523e8a407        hub.x.com/indexer:production                 "/home/app/init.sh"    4 minutes ago       Up 33 seconds                                                  indexer-9-worker                  
9ea54ce2f988        hub.x.com/indexer-manager                    "ruby /mech/bin/mech   5 minutes ago       Up 2 minutes                                                   indexer-5-manager                 
ca6de0a66087        hub.x.com/indexer:production                 "/home/app/init.sh"    5 minutes ago       Up 2 minutes                                                   indexer-3-worker                  
60aff3f205f8        hub.x.com/indexer-manager                    "ruby /mech/bin/mech   5 minutes ago       Up 2 minutes                                                   indexer-3-manager                 
eaf18b2f3c4e        hub.x.com/indexer-manager                    "ruby /mech/bin/mech   5 minutes ago       Up 3 minutes                                                   indexer-1-manager                 
a1eebaf620df        hub.x.com/indexer-manager                    "ruby /mech/bin/mech   29 minutes ago      Up 29 minutes                                                  indexer-9-manager                 
378bdb7c2709        hub.x.com/indexer-manager                    "ruby /mech/bin/mech   29 minutes ago      Up 29 minutes                                                  indexer-7-manager                 
90aaa9006ce0        hub.x.com/heartbeat:production               "/home/app/init.sh"    24 hours ago        Up 24 hours                                                    heartbeat-1-worker                
0ecb801976d6        hub.x.com/heartbeat-manager                  "ruby /mech/bin/mech   24 hours ago        Up 24 hours                                                    heartbeat-1-manager               
ec8d00d22d24        hub.x.com/metrics-frontend:production        "nginx -g 'daemon of   28 hours ago        Up 28 hours         80/tcp, 443/tcp                            metrics-frontend-1-worker         
173277fcc50a        hub.x.com/metrics-frontend-manager           "ruby /mech/bin/mech   28 hours ago        Up 28 hours                                                    metrics-frontend-1-manager        
94fe98f167b0        hub.x.com/metrics-backend:production         "/home/app/init.sh"    28 hours ago        Up 28 hours                                                    metrics-backend-1-worker          
b2ca34d7153a        hub.x.com/metrics-backend-manager            "ruby /mech/bin/mech   28 hours ago        Up 28 hours                                                    metrics-backend-1-manager         
7ba53e89252e        hub.x.com/notification-consumer:production   "/home/app/init.sh"    5 days ago          Up 5 days                                                      notification-consumer-1-worker    
5287792278e3        hub.x.com/notification-consumer-manager      "ruby /mech/bin/mech   5 days ago          Up 5 days                                                      notification-consumer-1-manager   
921a807fd2f9        hub.x.com/monitoring:production              "/home/app/init.sh"    6 days ago          Up 6 days                                                      monitoring-compute1-worker        
b59eac1c45d5        hub.x.com/monitoring-manager                 "ruby /mech/bin/mech   7 days ago          Up 7 days                                                      monitoring-compute1-manager       
47c1bc68ee58        hub.x.com/rails4test                         "/home/app/init.sh"    7 days ago          Up 7 days                                                      rails4test-1-worker               
17ed81a3c7de        hub.x.com/rails4test-manager                 "ruby /mech/bin/mech   7 days ago          Up 7 days                                                      rails4test-1-manager              
0511656673c8        hub.x.com/nginx-proxy                        "/app/init.sh"         10 days ago         Up 10 days          0.0.0.0:80->80/tcp, 0.0.0.0:443->443/tcp   nginx-proxy-1-worker              
fda97c877e35        hub.x.com/nginx-proxy-manager                "ruby /mech/bin/mech   10 days ago         Up 10 days                                                     nginx-proxy-1-manager             
b8e6910c5051        hub.x.com/scheduler                          "/home/app/init.sh"    2 weeks ago         Up 2 weeks                                                     scheduler-1-worker                
662714d472c6        hub.x.com/scheduler-manager                  "ruby /mech/bin/mech   2 weeks ago         Up 2 weeks                                                     scheduler-1-manager               
74685dc94f9d        mongo:3.0                                                  "/entrypoint.sh mong   2 weeks ago         Up 2 weeks          27017/tcp                                  mongo-main-mongos-1-worker        
b680d0ac7138        hub.x.com/mongo-main-mongos-manager          "ruby /mech/bin/mech   2 weeks ago         Up 2 weeks                                                     mongo-main-mongos-1-manager       
fafcfc43b620        mongo:3.0                                                  "/entrypoint.sh mong   2 weeks ago         Up 2 weeks          27017/tcp                                  mongo-blob-mongos-1-worker        
10ba331ff4dd        hub.x.com/mongo-blob-mongos-manager          "ruby /mech/bin/mech   2 weeks ago         Up 2 weeks                                                     mongo-blob-mongos-1-manager       
25e7a0b21857        mongo:3.0                                                  "/entrypoint.sh --re   2 weeks ago         Up 2 weeks          27017/tcp                                  mongo-blob-6-worker               
14220fba8f20        mongo:3.0                                                  "/entrypoint.sh --re   2 weeks ago         Up 2 weeks          27017/tcp                                  mongo-main-18-worker              
76205e2835e7        hub.x.com/mongo-blob-manager                 "ruby /mech/bin/mech   2 weeks ago         Up 2 weeks                                                     mongo-blob-6-manager              
8c5b71056873        mongo:3.0                                                  "/entrypoint.sh --re   2 weeks ago         Up 2 weeks          27017/tcp                                  mongo-main-15-worker              
522dfce52377        hub.x.com/mongo-main-manager                 "ruby /mech/bin/mech   2 weeks ago         Up 2 weeks                                                     mongo-main-18-manager             
20399343a3ab        mongo:3.0                                                  "/entrypoint.sh --re   2 weeks ago         Up 2 weeks          27017/tcp                                  mongo-main-12-worker              
dd2b9b4337d8        hub.x.com/mongo-main-manager                 "ruby /mech/bin/mech   2 weeks ago         Up 2 weeks                                                     mongo-main-15-manager             
c21c2cc95b0c        hub.x.com/mongo-main-manager                 "ruby /mech/bin/mech   2 weeks ago         Up 2 weeks                                                     mongo-main-12-manager             
cc2decd0adcc        hub.x.com/mailer                             "/home/app/init.sh"    3 weeks ago         Up 3 weeks                                                     mailer-1-worker                   
1cfc49a1508d        hub.x.com/mailer-manager                     "ruby /mech/bin/mech   3 weeks ago         Up 3 weeks                                                     mailer-1-manager                  
a5e51a5a34c9        hub.x.com/marketing-website                  "/bin/passenger_init   3 weeks ago         Up 3 weeks                                                     marketing-website-1-worker        
b1afad57be02        hub.x.com/marketing-website-manager          "ruby /mech/bin/mech   3 weeks ago         Up 3 weeks                                                     marketing-website-1-manager       
106bff58eb3e        hub.x.com/carbon-writer                      "/bin/sh -c /root/in   5 weeks ago         Up 5 weeks                                                     carbon-writer-1-worker            
363e3cad02d5        hub.x.com/carbon-writer-manager              "ruby /mech/bin/mech   5 weeks ago         Up 5 weeks                                                     carbon-writer-1-manager           
9193ee500fd0        hub.x.com/orchestrator                       "/app/orchestrator.r   5 weeks ago         Up 5 weeks                                                     orchestrator                      

real	3m48.266s
user	0m0.012s
sys	0m0.007s
```

# Describe the results you expected:
Normally docker ps executes with no significant delay.

# Provide additional info you think is important:
We ran into this on our cluster. Is there any thing I could run to help you guys reproduce/trace the problem? We happened upon this error when one of our containers hit a bug and started being restarted a whole bunch of times.

**Comment by hermansc** [comment]
Do `docker ps -a`, you may have a large amount of exited containers. Cleaning up with:

```
$ docker rm $(docker ps -qa -f status=exited)
$ docker rmi $(docker images -q -f dangling=true)
```

made `docker ps` speedup for me when I had the same issue.

**Comment by phemmer** [comment]
The `docker info` command shows the number of containers:

> root@us-compute-1 ~ # docker -D info
Containers: 46

**Comment by thaJeztah** [comment]
@tinco is this still an issue?

**Comment by tinco** [comment]
Well it usually isn't but coincidentally it happened again last week and I think we found a good correlation. We have configured our containers to log to syslog, and our syslog forwards to an external host (via rsyslog). Unfortunately this process is blocking (we have yet to configure it better) and when some node goes bananas it'll pump gigabytes of logs over it causing it to be throttled. We have found that during this throttling docker logs will be minutes behind and at the same time docker sometimes becomes unresponsive.

What I think might be happening: A docker container is logging but is throttled to the point where it has minutes of buffered logs. We try to kill it by issueing docker stop. Docker wants to do it but it wants to flush the logs first and gets stuck on that. Docker becomes unresponsive.

I don't know enough of Dockers internals to know for sure, just a theory.

**Comment by thaJeztah** [comment]
Sounds plausible; docker trying to send logs, causing the delay. Thanks for the additional info

**Comment by tinco** [comment]
Hi guys, I just confirmed my theory I think. It happened again, the direct solution was to find the pid of rsyslogd, it didn't respond to kill, so I did kill -9 on rsyslogd, and immediately docker started working again.

**Comment by thaJeztah** [comment]
Thanks @tinco!

@cpuguy83 @LK4D4 any idea if we can make modifications in docker to prevent this situation / making it not block because of this?

**Comment by cpuguy83** [comment]
I think it's plausible that this is fixed in 1.10.

In <= 1.9 the buffer would grow forever until it consumed all memory.
In 1.10 the buffer is capped (and also completely re-written)

**Comment by tinco** [comment]
We have planned a migration to Docker 1.10 somewhere in the coming month. Unfortunately we can only reproduce this when our cluster is in a dire state, so I think we can close this for now and I will reopen it if it happens again under Docker 1.10+.
